### PR TITLE
Add FreeDNS plugin support to acme.sh

### PR DESCRIPTION
**Pull request by dkerr64**
New pull request to replace pull request 557 with a clean branch without all the commits to clean up the code

### CODE REVIEW

**Comment by Neilpang** [comment]
Please update the readme.md files with the usage of freedns.

**Comment by Neilpang** [comment]
Please fix the error:  https://travis-ci.org/Neilpang/acme.sh/jobs/197461974

File: README.md [hunk 1]
```diff
@@ -270,6 +270,7 @@ You don't have to do anything manually!
 1. ISPConfig 3.1 API
 1. Alwaysdata.com API
 1. Linode.com API
+1. FreeDNS
```
Comment (Neilpang) [inline README.md]: change to:
```
FreeDNS(https://freedns.afraid.org/)
```

File: dnsapi/README.md [hunk 2]
```diff
@@ -278,6 +278,32 @@ acme.sh --issue --dns dns_linode --dnssleep 900 -d example.com -d www.example.co
...
+reuses that when needed.
+
+Now you can issue a certificate.
+
+```sh
+acme.sh --issue --dns dns_freedns --dnssleep 30 -d example.com -d www.example.com
+```
+
+FreeDNS updates records quite quickly so it is possible to reduce the dnssleep time, in the above example
+to 30 seconds.
```
Comment (Neilpang) [inline dnsapi/README.md]: I don't think explicitly using `30` seconds is a good idea.
I would prefer to use the default value.

File: README.md (hunk 1 above)
Comment (dkerr64) [inline README.md]: Done

File: dnsapi/README.md (hunk 2 above)
Comment (dkerr64) [inline dnsapi/README.md]: Done

**Comment by Neilpang** [comment]
I just tried it, but It failed to add txt record.
can you please try again ?

**Comment by Neilpang** [comment]
Another question,  why you save the cookies, instead of saving the username and password ?
How long will the cookies be valid ?

**Comment by dkerr64** [comment]
It is working for me.... just tried twice now.  Was any error message issued?  if you run with --debug 2 then all HTML returned will be printed out too.

As for cookie... according to Joshua Anderson (admin for FreeDNS whom I consulted when creating this) this value will change only when password changes.  He recommended storing that and not the userid/password.... it is not necessary to login and obtain the cookie unless the password changes.

David

**Comment by dkerr64** [comment]
Thoughts on why it might fail for you... I'm guessing that you are in china.  It is possible that the locale / language may be detected by the freeDNS website and different HTML returned.  It may be necessary to force to english/usa in the HTTP header sent.  Just a thought... would be helpful if you could determine if this is right.

David

**Comment by Neilpang** [comment]
Please use a new domain to test. I guest you have cache.

**Comment by dkerr64** [comment]
Are you testing with one of FreeDNS public domains or your own private domain?  FreeDNS public domains will not work for various reasons... 1) rate limiting at LetsEncrypt (there is much discussion of this at LetsEncrypt community pages) and 2) FreeDNS uses a captcha image for public domains to prevent bots creating lots of subdomains under free public domains.

I have added a test for failures to add TXT record because of the captcha/security code and issue an error message advising that public domains will not work.  And I have updated the README.md

If you are failing with a public domain please run with --debug 2 and send me the output. 

And yes, I am testing with new domains, no caching going on.

I have also added Accept-Language:en-US to the HTTP headers as a precaution agains future multi-lingual FreeDNS pages.

David

**Comment by Neilpang** [comment]
Yes,  I was testing with the public free domains.

**Comment by MukilanP** [comment]
@Neilpang

How did you skip the captcha issue while testing with public free domains. 

hu Dec 22 13:48:37 STD 2022] _ret='0'
[Thu Dec 22 13:48:37 STD 2022] FreeDNS failed to add TXT record for _acme-challenge.acmetest as FreeDNS requested security code
[Thu Dec 22 13:48:37 STD 2022] Note that you cannot use automatic DNS validation for FreeDNS public domains

**Pull request merged**